src/include/duckdb/execution/operator/join/physical_range_join.hpp

```diff
@@ -50,45 +49,85 @@ class PhysicalRangeJoin : public PhysicalComparisonJoin {
 
 	class GlobalSortedTable {
 	public:
-		GlobalSortedTable(ClientContext &context, const vector<BoundOrderByNode> &orders, RowLayout &payload_layout,
-		                  const PhysicalOperator &op);
+		GlobalSortedTable(ClientContext &client, const vector<BoundOrderByNode> &orders,
+		                  const vector<LogicalType> &payload_layout, const PhysicalRangeJoin &op);
 
 		inline idx_t Count() const {
 			return count;
 		}
 
 		inline idx_t BlockCount() const {
-			if (global_sort_state.sorted_blocks.empty()) {
-				return 0;
-			}
-			D_ASSERT(global_sort_state.sorted_blocks.size() == 1);
-			return global_sort_state.sorted_blocks[0]->radix_sorting_data.size();
+			return sorted->key_data->ChunkCount();
+		}
+
+		inline idx_t BlockStart(idx_t i) const {
+			return MinValue<idx_t>(i * STANDARD_VECTOR_SIZE, count);
+		}
+
+		inline idx_t BlockEnd(idx_t i) const {
+			return BlockStart(i + 1) - 1;
 		}
 
 		inline idx_t BlockSize(idx_t i) const {
-			return global_sort_state.sorted_blocks[0]->radix_sorting_data[i]->count;
+			return i < BlockCount() ? MinValue<idx_t>(STANDARD_VECTOR_SIZE, count - BlockStart(i)) : 0;
+		}
+
+		inline SortKeyType GetSortKeyType() const {
+			return sorted->key_data->GetLayout().GetSortKeyType();
 		}
 
-		void Combine(LocalSortedTable &ltable);
 		void IntializeMatches();
+
+		//! Combine local states
+		void Combine(ExecutionContext &context, LocalSortedTable &ltable);
+		//! Prepare for sorting.
+		void Finalize(ClientContext &client, InterruptState &interrupt);
+		//! Schedules the materialisation process.
+		void Materialize(Pipeline &pipeline, Event &event);
+		//! Single-threaded materialisation.
+		void Materialize(ExecutionContext &context, InterruptState &interrupt);
+		//! Materialize an empty sorted run.
+		void MaterializeEmpty(ClientContext &client);
+		//! Print the table to the console
 		void Print();
 
-		//! Starts the sorting process.
-		void Finalize(Pipeline &pipeline, Event &event);
-		//! Schedules tasks to merge sort the current child's data during a Finalize phase
-		void ScheduleMergeTasks(Pipeline &pipeline, Event &event);
+		//! Create an iteration state
+		unique_ptr<ExternalBlockIteratorState> CreateIteratorState() {
+			auto state = make_uniq<ExternalBlockIteratorState>(*sorted->key_data, sorted->payload_data.get());
+
+			// Unless we do this, we will only get values from the first chunk
+			state->SetKeepPinned(true);
+			state->SetPinPayload(true);
```

Comment:
From what I'm seeing right now, these created iterator states have quite a long lifespan. Doing this will cause all blocks you ever iterate over to stay pinned until the iterator state is destroyed, which can cause OOM exceptions even though we could've unpinned.

Calling `SetKeepPinned` again will cause the held pins to be dropped (regardless of whether the argument is `true` or `false`). Once a chunk has been emitted, the corresponding pins can be dropped, so this should be called once per emitted chunk.

Reply:
they are used in a buh of places with different lifetimes and memory footprints. The IEJoin internals should keep everything pinned and the parallel pair logic should limit the memory footprint. I'd like to have a better algorithm for choosing parallel block sizes based on memory footprint. Right now its hard-coded to 1024 vectors.